Further tidy up of api link parsing and testing

SpiritLevel opened the pull request.

dhensby commented on tests/DocumentationParserTest.php:
```diff
@@ -325,21 +320,21 @@ public function testApiLinks()
 
         // expected url format resulting from rewriting api shortcode links
-        $url_format = '<a href="http://api.silverstripe.org/search/lookup/?q=%s&version='.$page_version.'&module=documentationparsertest">%s</a>';
+        $html_format = '&nbsp;<a href="http://api.silverstripe.org/search/lookup/?q=%s&version='.$page_version.'&module=documentationparsertest">%s</a>';
```
Why do we need the nbsp?

SpiritLevel replied:
There is something strange going on that I haven't quite pinned down: the first and last characters sometimes get gobbled up. The &nbsp; puts a space back in front. And, in the instructions, I say that one must put a space at the end of the shortcode. This is not satisfactory but works but it works most of the time. I am trying to get backticks to work with the api links like this `[title](api:link)` as they do with ordinary markdown links of the form `[title)(link)`. I'll try to get this sorted...

dhensby replied:
Why does this need to happen now and didn't before?

SpiritLevel replied:
It wasn't "needed" before but neither was it working correctly before!

Anyhow, I think I've finally sorted the gobble problem :) Bloody regexes are tricky to get exactly right...Ready to merge.

dhensby commented:
Fantastic!

SpiritLevel commented:
Yay...merged! I noticed there are similar problems with relative links and image links so was thinking of fixing these up too....api links are [title](api:link) or [api:link] whereas relative links are [title](link) or [link] and images are ![](link). Relative links are too similar to ordinary markdown links and the ![] for images is not nice. I was thinking of standardising:

[title](api:link) or [api:link] for api links

[title](doc:link) or [doc:link]  for relative links

[img:link] for image links

[rev:date] for changelog revisions

[sec:...] for security

But, this means people writing documentation would have to adopt the new syntax. And,  I would have to do a fairly quick search and replace on all docs to change to the new syntax.

dhensby commented:
![] for images is standard markdown so we shouldn't stop diking they for our own proprietary image syntax

SpiritLevel commented:
What about using  both `[img:link]` and `![]` ? The goal of markdown is to be more readable, and I think `[img:link]` is the more readable. And, just as with relative and api links, custom behaviour has already been added on top of the underlying markup for `![]`. Perhaps leave `![]` exactly as in markdown and put the customisations into `[img:link]` ? What about the other suggestions, like doc:, for the same reasons ?

camfindlay commented:
Should probably avoid making our own standards for things unless we absolutely have to. eg. framework will be moving towards PSR-2 as it's better for us to adopt existing standards.

SpiritLevel commented:
Well, DocumentationParser already changes a few things in standard markdown. It changes how []() and [] work so that relative links are handled in a special way. And it changes how ![] works for images stored in an _images directory. So, SS has already altered the behaviour of the underlying markdown. My suggestion is to undo this: change DocumentationParser so that markdown is left in its untouched original form and then introduce special SS documentation syntax with doc:, api:, img:, rev:, sec: as additions to standard markdown rather than making changes to existing standard markdown. This way, using standard markdown in SS docs is the same as using it outside the SS docs. But, additional syntax is available in SS to refer to the api via absolute url, and to images and md files stored in documentation directories via relative urls. To me, it makes a lot of sense to organise things this way.

camfindlay commented:
sure, can you explain what each of your new tags do please. Could we also perhaps start with only the ones we actually need rather than try preempting them eg. why do we need img: sec: and rev:?

SpiritLevel commented:
sec: and rev: were just ideas I threw into the mix since, while going through the docs, I noticed a lot of inconsistencies in how revisions and security updates are rendered. I envision these tags would be used something like this in the docs:

`[rev:2015-12-15](description of the revision)`

`[sec:2015-12-15](description of the security release)`

Nice and standard sytnax :) We could choose to have these rendered as any html we like, which is an advantage over using only the underlying standard markdown.

The api: tag needed fixing since it wasn't properly constructing the api url query for methods and properties. Also, they couldn't be backticked (ie: commented out). I have fixed this, although with the issue Dan raised. To see what the api: tags do, read the comments at the start of DocumentationParser::rewrite_api_links(). There are 6 cases: classes, properties, and methods with titles and without titles. This is a custom tag that adds to markdown, while not altering it.

If you look at DocumentationParser::rewrite_relative_links() and DocumentationParser::rewrite_image_links() you will see that the standard behaviour of `[]()` and `[]` and `![]` of the underlying markdown are being altered. 

I am suggesting that DocumentationParser::rewrite_relative_links() and DocumentationParser::rewrite_image_links(), and perhaps other methods, be changed so that they leave `[]()` and `[]` and `![]` alone and instead only parse links with img: and doc: in them. Then we can choose to keep how they are rendered in html or change the rendering, if we wish to. The only definite change will be in how the links are actually written in the docs, ie, one will have to use img: and doc: tags. Most importantly, the underlying standard markdown will remain in its original state, not altered in any way by DocumentationParser as is the current situation.

Have a look at https://docs.silverstripe.org/en/3.0/misc/ss-markdown/ to see the original intention of SS markdown. 

Bottom line: I think that "SS markdown" should be a superset of standard markdown with its own distinguishing tags and that the underlying standard markdown works as expected. Hence, the syntax used in writing docs needs to be clarified and fixed where broken. What do you all think ?

camfindlay commented:
Can we start with api: get that working as intended so we can move this forward, then come back and look to implement others at a later point?

dhensby commented:
I think having some kind of extension to markdown that helps us insert links dynamically is what we're after.

This can be done by extending Parsedown and that should be our approach

camfindlay commented:
+1 I'd support an extension of the parser as an approach

SpiritLevel commented:
@dhensby @camfindlay 

Connected to https://github.com/silverstripe/silverstripe-docsviewer/pull/94

OK, are you guys suggesting subclassing Extension to extend ParsedownExtra (or Parsedown itself?) as in https://docs.silverstripe.org/en/3.2/developer_guides/extending/extensions/? Or, are you thinking of directly subclassing ParsedownExtra (or Parsedown) ? I'm guessing the former :)

However, even with an extension, there are still the issues of (i) the algorithm to parse api: tags and (ii) what sort of documentation syntax we want, ie, what should "SS Markdown" have in it ? Regarding (i), have a look at latest commit: https://github.com/SpiritLevel/silverstripe-docsviewer/commit/e443e9fa0b93012949be2a1fa4d719a6235f32ed

Reading between the lines a little, @dhensby, I get the sense that there is some higher level re-factoring you'd like to see done on docsviewer ? I have had that urge too. The tests, and how versions are handled, seem a bit disorganised. Things feel, generally, more complex than necessary. A re-factor might also be helpful for the big SS4 release? I'll do the hard yakka if you lead me through.

At the same time, it seems sensible to consider what is needed for userhelp.ss.org; IMVHO (V=very!), I think the userhelp docs should be in the docs folder in the cms repo rather than in silverstripe-userhelp-content.

dhensby commented:
> OK, are you guys suggesting subclassing Extension to extend ParsedownExtra (or Parsedown itself?) as in https://docs.silverstripe.org/en/3.2/developer_guides/extending/extensions/? Or, are you thinking of directly subclassing ParsedownExtra (or Parsedown) ? I'm guessing the former :)

Directly extending Parsedown(Extra) which seems to be the way to add behaviour to that class. At the moment we have disconnected functions that aren't parsing the markdown the same way that the main body is.

Yes, we need to have an agreed addition to the markdown standard for our own needs and I agree with @camfindlay where he says start with `api:` and look at others later.

Yes, some form of refactor is needed; I think because we weren't using parsedown until recently... But now we are we should extend it properly and use it's internal parsing to add our rules, rather than this rather nasty mangling of the markdown before parsing

dhensby commented:
> At the same time, it seems sensible to consider what is needed for userhelp.ss.org; IMVHO (V=very!), I think the userhelp docs should be in the docs folder in the cms repo rather than in silverstripe-userhelp-content.

This is a bit off topic, IMO - let's leave it for now

SpiritLevel commented:
OK :) Agreed. Let's do api: first.

Just to clarify, you are suggesting:

`class SSParsedown extends ParsedownExtra {...}`

Before committing to Parsedown(Extra) take a look at this:

https://help.github.com/articles/updating-your-markdown-processor-to-kramdown/

I am wondering if a bit of thought needs to go into choosing the best Markdown processor for SS's needs. I did a quick search and there appear to be ports of kramdown to PHP. Or, do you think it is best to stay with Parsedown(Extra) ?

dhensby commented:
I'm suggesting the extension, yes, as that appears to be how you're meant to extend Parsedown.

I don't think we should use kramdown as it's a ruby implementation which adds another layer to the stack and would be harder for us to customise and maintain.

SpiritLevel commented:
Lots of PHP implementations of Markdown:

https://github.com/markdown/markdown.github.com/wiki/Implementations

SpiritLevel commented:
Looks like Parsedown might be the one to go with, not Parsedown Extra....not yet, anyway.

dhensby commented:
`Parsedown` seems to be a pretty fast one and we currently use `ParsedownExtra`, which is just an extension to `Parsedown`, so I don't see a reason not to use it.

SpiritLevel commented:
Yeah...I noticed that Parsedown and ParsedownExtra were quite a bit faster than Markdown and MarkdownExtra. I also noticed that ParsedownExtra is not as active and the build is currently failing....I suppose the best thing to do is examine what, if anything, is being used or is needed from ParsedownExtra. If nothing, perhaps it would be better to just use Parsedown as the base class for SSParsedown ?

dhensby commented:
If we don't need the extra stuff, then sure

SpiritLevel commented:
OK, I will start taking a look at Parsedown and ParsedownExtra in more detail :)

dhensby commented:
`ParsedownExtra` should give some indication of how you can add custom syntax to the parser

SpiritLevel commented:
Yup...I'll use that as a model :)
